### PR TITLE
FB-123 Introduce a caching system

### PR DESCRIPTION
**Pull request by zazabe**

### CODE REVIEW

**Comment by zazabe** [comment]
@njam please review

**Comment by njam** [comment]
Some notes about the skeleton:

- README:
  - Add instructions how to release a new version (see [guides](https://github.com/cargomedia/guides/tree/master/software-projects#documentation)
  - Add NPM badge ([example README](https://raw.githubusercontent.com/cargomedia/loggly-github/master/README.md))
  - Describe how the request can be sent to the daemon, maybe add an example?
- bin/cmd.js: call this file `cm-bundler`?
- bin/cmd.js: Rename `rid` to `requestId`
- bin/cmd.js: Maybe move most of the business logic into a `server.js` to keep the binary small
- bin/cmd.js: Should this `moreVerbose` maybe just set the logger level to "silly" - https://github.com/winstonjs/winston#logging-levels ?
- bin/cmd.js: what about unexpected errors? Is it okay that they make the program crash? E.g. I can crash it with `echo -n '{"config":1,"command":"code"}\u0004' | nc -U /tmp/socket` (ZSH).
- config: I've found it valuable to validate the config, maybe this [example using "json schema"](https://github.com/cargomedia/loggly-github/blob/dddc60b423661aa3cdf9e77b7f10a5678a12d47d/src/config.js) is interesting, could also be stored in a file. http://json-schema.org/
- socket/server.js: check error handling in `_onData()`. should this errors be logged?

Let's discuss, and I will have a look at the real code tomorrow.

**Comment by njam** [comment]
- config.js / cmd.js: Default host `0.0.0.0` might be a bit dangerous, as this daemon potentially can be used to read files on the system. How about `127.0.0.1`?
- lib/stream/filter: why push `null` to the stream? What does it do?
- Did you see this mississippi? https://github.com/maxogden/mississippi
- bundler/config/cache: The `timeout` is used only for `BundlerConfig.initialize()` (not for `_fillCache()`) - on purpose? If yes, then probably no need to make it configurable?
- bundler/plugins/*: Add a docu to these files about what they do?
- bundler/plugins/content: unused var "noParse"

Looks great!

**Comment by zazabe** [comment]
> lib/stream/filter: why push null to the stream? What does it do?

It indicate the end of the stream, see https://nodejs.org/api/stream.html#stream_readable_push_chunk_encoding

> Did you see this mississippi? 

Yes, but it's actually more like [an aggregation of modules](https://github.com/maxogden/mississippi/blob/master/index.js), I thought it's better to decide which module we want to include by ourself, because we use few of them only.
There's also [mississippi2](https://www.npmjs.com/package/mississippi2), which is [not so popular yet](https://npmcompare.com/compare/mississippi,mississippi2) compared to mississippi...


> bundler/config/cache: The timeout is used only for BundlerConfig.initialize() (not for _fillCache()) - on purpose? If yes, then probably no need to make it configurable?

Yes, it's the max time to create + watch a new config... it was maybe useful when the watch was on NFS... config removed!

**Comment by zazabe** [comment]
@njam please check again

File: README.md [hunk 1]
```diff
@@ -17,32 +18,101 @@ Usage
...
+
+
+##### example
+```bash
+echo -ne '{"name":"test","command":"code","config":{}}\004' | nc 127.0.0.1 6644 
+```
+
 #### JSON configuration
 
 ```js
```
Comment (njam) [inline README.md]: This JSON block is not rendered so nicely. All comments are red (errors). Maybe better use no type at all?

File: config.prod.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    enabled: false
+  },
+  uglify: {
+    enabled: true
+  },
+  cache: {
+    config: {
+      max: 30,
+      maxAge: 5 * 60 * 1000,
+      timeout: 5000
```
Comment (njam) [inline config.prod.js]: deprecated?

File: README.md [hunk 3]
```diff
@@ -17,32 +18,101 @@ Usage
...
+    output: {
+      quote_keys: true,
+      beautify: false
+    }
+  },
+  cache: {                              // see https://github.com/isaacs/node-lru-cache
+    config: {
+      max: 12,
+      maxAge: 7 * 24 * 60 * 60 * 1000,
+      timeout: 5000                     // max time to create + watch a config
```
Comment (njam) [inline README.md]: deprecated?

**Pull request closed without merging**